i #1 Created event_processing file with 3 methods to run analysis

connorn-dev opened the pull request:
Methods: 
- start_end_activities(csv_path)
- generate_tree(csv_path, output_dir)
- generate_graph(csv_path, output_dir)

Notes: Installs pm4py if not already installed.

connorn-dev committed:
Created event_processing file with 3 methods to run analysis on given csv file

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

connorn-dev committed:
i #1 added comments to name .png file

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

connorn-dev committed:
i #341 Updated script to be called from terminal

- Added config_file and action argument
- Run chosen action

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

connorn-dev committed:
i #1 Created Python Notebook with example

- Added Python notebook with all functionality.

connorn-dev committed:
i #1 Deleted script and changed notebook name

- Deleted .py script to have notebook replace
- Changed name of notebook

connorn-dev committed:
i #1 added examples section and more models

The following was added:
- Example section at bottom of notebook
- Petri Net model
- Performance Models
- BPMN filtered model
- Cleaned up wording
- Images folder

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

carlosparadis commented on images/occurrence_dfg.png:
Do not version binary files. Generate them on the fly in the code so users generate it themselves when they run your notebook.

carlosparadis commented on images/performance_dfg.png:
Do not version binary files. Generate them on the fly in the code so users generate it themselves when they run your notebook.

carlosparadis commented on images/petri_net.png:
Do not version binary files. Generate them on the fly in the code so users generate it themselves when they run your notebook.

carlosparadis commented on images/process_graph.png:
Do not version binary files. Generate them on the fly in the code so users generate it themselves when they run your notebook.

carlosparadis commented on images/process_tree.png:
Do not version binary files. Generate them on the fly in the code so users generate it themselves when they run your notebook.

carlosparadis commented on images/starting_data.png:
Do not version binary files. Generate them on the fly in the code so users generate it themselves when they run your notebook.

carlosparadis commented on issue_event_processing.html:
Should also be removed, you want to version the notebook, not the html.

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+   "metadata": {},
+   "source": [
+    "# Introduction\n",
+    "Process mining helps uncover insights from data that represent different events in a process. It is valuable for understanding workflows, finding inefficiencies, and optimizing development processes. Using Kaiaulu  for process mining enables this discovery for Github Events (e.g. creating issues, push events, deleting branches, etc). Allowing users to visualize and analyze event data effecti ...
```
The writing here is a bit too flourished: 

> helps uncover insights

> valuable for understanding

> finding inefficiencies

> optimizing development 

> analyze event data effectively

This does not contribute anything factual to the reader and it is not why they are interested in looking for this Notebook. 

What you want to start here is by explaining what *IS* event data, then proceed to say what process mining factually adds to it. 

For the GitHub Event data, you want to briefly and explain reference the Kaiaulu Notebook (reference the master branch notebook in Kaiaulu) for details. What should you say here? That the GitHub Event endpoint tracks interactions between user and github (such as whichever events you have). 

Right after, you want to do the subroutine call to the event data, and display the table to the user. Explain a bit and show what the data looks like.

Next, you motivate your process mining portion: Say that seeing from the raw table whether users are all following the same process or not from the raw table is not easy. That motivates experiment 1 (please dont name the section that, name it after what you are about to show).

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+   "source": [
+    "# Save and set the current working directory to Kaiaulu\n",
+    "cwd = os.getcwd()\n",
+    "os.chdir(os.path.expanduser(\"~/Desktop/Kaiaulu/Working_issues/kaiaulu/\"))\n",
```
Use relative paths, explicitly create a code block that says output and tell the user by default it will save to the same folder this code was downloaded (the only path you know for certain will exist on user side). 

You can also assume the Kaiaulu recommended path structure with raw_data folder exists.

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+   "metadata": {},
+   "source": [
+    "### Download and Parse data with Kaiaulu ghevents.R (CLI)\n",
+    "\n",
```
You need to walk the user here. Assume I just gave you this Notebook on your day 1. Would it suffice for you to understand what is going on? Explain briefly that Kaiaulu is an R package that defines various functions in an API. The API is used in the Notebook and in executable scripts. Then say for details on the executable, the user can read the notebook since the underlying functions are the same. Nonetheless, you should make your notebook self contained.

So you have to re-explain here minimally what the executable will give and what is this ssh token (link the user to the docs page of github and briefly explain it).

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Download and Parse data with Kaiaulu ghevents.R (CLI)\n",
```
rename to: GitHub Events

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+    "cwd = os.getcwd()\n",
+    "os.chdir(os.path.expanduser(\"~/Desktop/Kaiaulu/Working_issues/kaiaulu/\"))\n",
+    "\n",
+    "# To download use the download command specifing the <config_file> <github_token>\n",
```
This is too many sub-sections. You want to be consistent that the section division doesn't get too specific or it defeats the purpose of being a section. 

Change to be a sub-section, enumerate the section (it is hard on the eye to see what is header or sub-header when note close together, so use 1., 1.1, etc). Rename to: GitHub Events using Kaiaulu

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+    "\n",
+    "# To download use the download command specifing the <config_file> <github_token>\n",
+    "\n",
+    "command = f\"Rscript exec/ghevents.R download conf/kaiaulu.yml --token_path=~/.ssh/github_token\"\n",
```
This is not a good path. It assumes the user is running inside the Kaiaulu project the exec. Rather, the reasonable assumption is that the user downloaded your code (process_mining) and wants to run the exec from Kaiaulu. So the current working directory should assume the user is at process_mining folder, goes one level above, then access the kaiaulu folder. 

What I recommend you do is to explain the expected folder organization of the projects before throwing paths left and right. Assume the following:

kaiaulu/kaiaulu/exec/ghevents.R
kaiaulu/rawdata/
process_mining/notebooks/issue_event_processing.ipynb

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+    "\n",
+    "As stated above it is reccomended you start with only a few event issues. To do this you can open the created issue_output.csv with Excel or Google Sheets and modify the table to only include a few. \n",
+    "\n",
+    "Note: commit_output.csv has been implemented for furture development and is note currently used for any process modeling. "
```
I have no idea what is going on here. These should not be the file names that Kaiaulu will download the rawdata as. How do you go from the rawdata Kaiaulu downoaded to these csvs? Why are these two tables being passed as input here? Are both being generated by ghevents download? That sounds odd.

carlosparadis commented on issue_event_processing.ipynb:
```diff
@@ -0,0 +1,458 @@
+   "source": [
+    "#### Execute Parse Command\n",
+    "\n",
+    "As stated above it is reccomended you start with only a few event issues. To do this you can open the created issue_output.csv with Excel or Google Sheets and modify the table to only include a few. \n",
```
This is not useful. The user will not be able to create the files by hand with just this as instruction. Did you not say you created the fake generator for the examples? Where are they located? That's what they were for here.

carlosparadis reviewed (changes requested):
Hi Connor,

This needs substantial revision. I did not review the entirety of the notebook because the narrative is somewhat confusing.

The overall structure of the notebook should be something like this:

- You introduce the user to GitHub Events. You explain you can obtain them with Kaiaulu. You motivate your code, saying you are interested in seeing if developers often follow a similar process. This is all brief, the details go in subsequent sections.
- Right after the introduction, you explain the overall folder organization between kaiaulu, process mining, and the raw data exec will give to you. 
- GitHub Events: You minimally explain that Kaiaulu organizes its features in functions, which are used in Notebooks and execs. For details on how to construct config file for events, you defer the user to reading the Events Notebook (do ensure the notebook does explain what you claim it does). 
     - Explain that by running the config, Kaiaulu will save the files to rawdata (specify the exact folder). Don't create a sub-section for this. 
     - After executing the script, simply load the table in pandas and show the user what it looks like. Explain to the user.  Again no need to keep making sub-sections here, GitHub Events section suffice. 
     - Proceed to explain and re-motivate your work, that is hard to guess if people are following the same process. 
- Process Mining GitHub Events: 
   - In this new section, begin by explaining you will demonstrate it in practice. Using the *already loaded pandas dataframe* of the real dataset, apply the 3 process mining functions. 

By this point, the user should be able to use your code to their needs. What you then need is a section that explains experiments 1, 2 and 3. Where are the fake data generators here? 

Also, where is the api folder?  Where is the markdown? This does not look like the most recent version of the code. I still see functions being defined in the notebook.

Also: 

Where is the env.yml conda file?

connorn-dev committed:
Separated Notebook from API folder & added pdoc support

This version includes the following:
- A separate API folder with process mining functions and helper functions
- Function comments added to support pdoc format

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

connorn-dev committed:
i #1 Changed comment

- Changed modify_event_in_csv function comment

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>

carlosparadis commented on api/csv_generator.py:
```diff
@@ -0,0 +1,78 @@
+fake = Faker()
+
+# Function to generate fake data
+def generate_csv_file(num_issues=1, num_events_per_issue=7, output_csv="generated_csv", seed=2):
```
Notice the name of your function vs the title of the docs: The more appropriate name here is generate_fake_event_log. The genetate_csv_file is too broad and any logic could be doing this.

carlosparadis commented on api/csv_generator.py:
```diff
@@ -0,0 +1,78 @@
+import random
+import pandas as pd
+from faker import Faker
+from datetime import datetime, timedelta
```
I believe standard practice in Python is to define __init__.py and place the imports there, but I may be mistaken. Please double check.

carlosparadis commented on api/csv_generator.py:
```diff
@@ -0,0 +1,78 @@
+from datetime import datetime, timedelta
+
+# Initialize faker
+fake = Faker()
```
API files should not be initializing objects. This is function definition only.

carlosparadis commented on api/csv_generator.py:
```diff
@@ -0,0 +1,78 @@
+import random
+import pandas as pd
+from faker import Faker
```
Nice finding!

carlosparadis commented on api/csv_generator.py:
```diff
@@ -0,0 +1,78 @@
+# Function to generate fake data
+def generate_csv_file(num_issues=1, num_events_per_issue=7, output_csv="generated_csv", seed=2):
+    """
+    Generates fake event log data and saves it to a CSV file.
```
Expand description so it explains what fake set of activities is happening. Given the poster feedback, you may want to do some adjustments to reflect a Kaiaulu workflow.

carlosparadis commented on api/csv_generator.py:
```diff
@@ -0,0 +1,78 @@
+    print(f"Fake event log data saved to {output_csv}")
+    return output_csv
+
+def modify_event_in_csv(csv_path, row_index, new_event, output_csv=None):
```
rename to `modify_event`, and rename first parameter to `event_log_df`. In memory, you have a pandas dataframe so you should refer to memory objects, except if the goal of the function is to save the csv itself. 

With that said, I am confused with the logic in this function. Why do you have a function that modifies csv when you can just manipulate everything in memory?

carlosparadis commented on api/io_helpers.py:
Just call it io.py

carlosparadis commented on api/io_helpers.py:
```diff
@@ -0,0 +1,45 @@
+import pm4py
+
+
+def generate_filename(prefix, extension="png"):
```
def generate_timestamped_filename

carlosparadis commented on api/io_helpers.py:
```diff
@@ -0,0 +1,45 @@
+    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
+    return f"{prefix}_{timestamp}.{extension}"
+
+def load_event_log(csv_path):
```
def read_event_log for consistency with pandas.read_csv

carlosparadis commented on api/process_visual_generation.py:
```diff
@@ -0,0 +1,245 @@
+
+## Features
+
+This module provides functions to:
```
Non-comment english in a .py file. Not sure this will work.

carlosparadis commented on api/process_visual_generation.py:
````diff
@@ -0,0 +1,245 @@
+
+## Example
+
+```python
````
Not sure what is going on here. Is this supposed to be a ipynb?

carlosparadis commented on api/process_visual_generation.py:
rename file to process_discovery.py ; stay consistent to the field terminology

carlosparadis commented on api/process_visual_generation.py:
```diff
@@ -0,0 +1,245 @@
+import pm4py
+
+
+def start_end_activities(csv_path):
```
Not sure this belongs to process_discovery.py. Seems event-log related? Maybe a file for event log manipulation may make more sense, or just place on io.py

carlosparadis commented on api/process_visual_generation.py:
```diff
@@ -0,0 +1,245 @@
+    """
+    Reads an event log from a CSV file and returns its start and end activities.
+
+    Assumes the CSV has the following columns:
```
Reference the python function that creates the .csv instead, and that function contains the specification of the csv returned. Have the user know what function to use, rather than explain it as if they will code it! Remember Kaiaulu has \code{\link{function_name}} in R. Find the equivalent in python. 

You should also see itm0 for the "See Also" section so you can reference on the returns what functions will use the output of another.

carlosparadis reviewed (changes requested):
Here some more comments on your API. I may need to do one more pass after all the changes to check the overall consistent of what file has what functions in the API, and how they are used on the Notebook after all the changes. 

Thanks!

connorn-dev committed:
i #1 Revised Notebook and Refactored API

- Notebook now starts the user from the beginning
- Uses relative paths
- Includes three "experiements"
- API now asks user action 'view' 'save' or 'both'
- Renamed API functions
- Includes env.yml

connorn-dev committed:
i #1 corrected paths and imports

- Corrected relative path and api import

carlosparadis commented on env.yml:
Please change this for me. What you did was export the env.yml from conda. This hardcoded the exact version down to commit hash of your environment, and also add here all the indirect dependency matrix of libraries you don't even use.

You want to create this file manually and only list on dependencies the libraries you actually import.

carlosparadis commented on env.yml:
```diff
@@ -0,0 +1,548 @@
+name: base
```
name it process_mining

carlosparadis commented on notebooks/issue_event_processing.ipynb:
```diff
@@ -0,0 +1,391 @@
+   "metadata": {},
+   "source": [
+    "# Introduction\n",
+    "Event logs are the foundation of process mining. They capture records of activities within a system, providing information about when actions occur and what those actions are. For example, in GitHub Issue Events, actions such as assigning users, labeling issues, and closing issues are recorded. Together, these events tell the full story of the process from start to finish. Event logs can be  ...
```
typo in diffrent. Please run a spell checker.

carlosparadis reviewed (changes requested):
Just need to patch the env file. The rest looks pretty good. Please check the env file of the sentiment project is also updated.

connorn-dev committed:
i #1 Updated env and finalized file paths

- env.yml changed
- file paths in notebook changed

connorn-dev committed:
i #1 full env added

- Added full env with versions

connorn-dev committed:
i #1 Finalized working env, fixed typo

Signed-off-by: Connor Narowetz <cnarowetz@gmail.com>
